esp.c

```diff
@@ -31,40 +31,93 @@
 #endif
 
 
-
 #ifdef HAS_ESP_IDF_5
 
+static SemaphoreHandle_t show_mutex = NULL;
+
 void espShow(uint8_t pin, uint8_t *pixels, uint32_t numBytes, boolean is800KHz) {
-  rmt_data_t led_data[numBytes * 8];
+  // Note: Because rmtPin is shared between all instances, we will
+  //  end up releasing/initializing the RMT channels each time we
+  //  invoke on different pins. This is probably ok, just not
+  //  efficient. led_data is shared between all instances but will
+  //  be allocated with enough space for the largest instance; data
+  //  is not used beyond the mutex lock so this should be fine.
+
+#define SEMAPHORE_TIMEOUT_MS 50
+
+  static rmt_data_t *led_data = NULL;
+  static uint32_t led_data_size = 0;
+  static int rmtPin = -1;
+
+  if (show_mutex && xSemaphoreTake(show_mutex, SEMAPHORE_TIMEOUT_MS / portTICK_PERIOD_MS) == pdTRUE) {
+    uint32_t requiredSize = numBytes * 8;
+    if (requiredSize > led_data_size) {
+      free(led_data);
+      if (led_data = (rmt_data_t *)malloc(requiredSize * sizeof(rmt_data_t))) {
+        led_data_size = requiredSize;
+      } else {
+        led_data_size = 0;
+      }
+    } else if (requiredSize == 0) {
+      // To release RMT resources (RMT channels and led_data), call
+      //  .updateLength(0) to set number of pixels/bytes to zero,
+      //  then call .show() to invoke this code and free resources.
+      free(led_data);
+      led_data = NULL;
+      if (rmtPin >= 0) {
+        rmtDeinit(rmtPin);
+        rmtPin = -1;
+      }
+      led_data_size = 0;
+    }
 
-  if (!rmtInit(pin, RMT_TX_MODE, RMT_MEM_NUM_BLOCKS_1, 10000000)) {
-    log_e("Failed to init RMT TX mode on pin %d", pin);
-    return;
-  }
+    if (led_data_size > 0 && requiredSize <= led_data_size) {
```

Comment:
[IMPORTANT]

`rmtInit()` does several things
- [if the pin configuration hasn't changed, do nothing](https://github.com/espressif/arduino-esp32/blob/1a42b87fb5a92e25596cff10ebe563a85b847fab/cores/esp32/esp32-hal-rmt.c#L471)
- [sets a callback that releases the RMT channel if the pin configuration is changed](https://github.com/espressif/arduino-esp32/blob/1a42b87fb5a92e25596cff10ebe563a85b847fab/cores/esp32/esp32-hal-rmt.c#L480)
- [resets the pin configuration to no peripheral](https://github.com/espressif/arduino-esp32/blob/1a42b87fb5a92e25596cff10ebe563a85b847fab/cores/esp32/esp32-hal-rmt.c#L497)
- [allocates an RMT channel](https://github.com/espressif/arduino-esp32/blob/1a42b87fb5a92e25596cff10ebe563a85b847fab/cores/esp32/esp32-hal-rmt.c#L547)
- [sets the pin to that RMT channel](https://github.com/espressif/arduino-esp32/blob/1a42b87fb5a92e25596cff10ebe563a85b847fab/cores/esp32/esp32-hal-rmt.c#L607)

`rmtDeinit()` just does one thing
- [resets the pin configuration if it was using RMT](https://github.com/espressif/arduino-esp32/blob/1a42b87fb5a92e25596cff10ebe563a85b847fab/cores/esp32/esp32-hal-rmt.c#L281)

Based on that, you can do just like previous code
- always call `rmtInit()`
- never call `rmtDeinit()`

That will keep RMT initialized when it should be, let it be de-initialized as needed (if the pin is repurposed with `pinMode()` or whatnot), and simplify the code a bunch. Assuming you agree (PLEASE CHECK MY LOGIC), you can
- remove `rmtPin` as a variable
- remove all the checks against `rmtPin` and calls to `rmtDeinit(rmtPin)`
- call `rmtInit(...)` unconditionally on entry, precisely as the previous code did

That would simplify this change to JUST changing the allocation semantics (with a mutex just in case), which should be a lot easier for everyone to review.

Reply:
This code handles more than just the caller changing the pin - since all instances of `Adafruit_NeoPixel` are sharing the RMT interface (and each instance defines its own output pin), this code handles re-initializing the RMT each time a different instance's `espShow()` is called. As noted in the comments, it's not very efficient but seems to work ok in practice.

The multi-threaded scenario I'm targeting is creating multiple instances of `Adafruit_NeoPixel` on the main thread, then launching individual threads to run each instance (and call `show()`). It would be nice to be able to handle the mutex initialization on the individual threads, but AFAICT there isn't a way to initialize the mutex that way without introducing a race condition.

Reply:
I see your point about being able to unconditionally call `rmtInit()` - I didn't dig into the Espressif implementation and treated it like a black box, assuming I would need to let the API know when the pin changed.

Reply:
Yeah, I think being "dumb" will be no less efficient (the underlying library already bails out in the no-change case), and that approach was blessed by the author of arduino-esp32 @me-no-dev in reviewing @ladyada's original change, and the "dumb" approach seems worth the rather large gain in simplicity and in keeping this change small. IMVHO

Reply:
So it looks like the way forward is to leave the `espInit()` logic in place, remove the conditional logic around `rmtInit()`, add some additional logging as indicated above, (and see if I can figure out how to clean up my whitespace issues 😜). Does that sound about right?

Its been a minute, so it's gonna take a bit for me to setup my environment (and hardware) and get back into the grove on this issue.

Reply:
Sounds good to me BUT we're a bunch of kibitzers guessing at what the Mama Ada would actually take, so if Ada folks jump in they should override any of our peanut-gallery comments. (OR they might just end up rolling their own fix??)